flink-runtime/src/main/java/org/apache/flink/runtime/deployment/InputChannelDeploymentDescriptor.java

```diff
@@ -52,30 +52,38 @@
 	/** The ID of the partition the input channel is going to consume. */
 	private final ResultPartitionID consumedPartitionId;
 
-	/** The location of the partition the input channel is going to consume. */
-	private final ResultPartitionLocation consumedPartitionLocation;
+	/** The location type of the partition the input channel is going to consume. */
+	private final LocationType locationType;
+
+	/** The connection to use to request the remote partition. */
+	private final Optional<ConnectionID> connectionId;
```

Comment:
I also considered using `ShuffleDeploymentDescriptor` here to replace `ConnectionID` before. But there are two concerns in implementation:

1.  In eager schedule mode when receiving all required slots, we might not assume the deployment sequence must be strict with topology sequence. That means the consumer execution deployment might be earlier than the producer execution. So in the process of `InputChannelDeploymentDescriptor#fromEdges`, we might not get cached SDD directly from producer execution. But we can generate `ConnectionID` based on other infos. Otherwise we must confirm the deployment sequence is from producer to consumer or generate producer's SDD during deploying consumer in `InputChannelDeploymentDescriptor#fromEdges`.

2. I thought of introducing `UnknownShuffleDeploymentDescriptor` before, but from semantic aspect it is a bit redundant with `LocationType.Unknown`. In addition, it seems no specific usages like ` instanceof UnknownShuffleDeploymentDescriptor` in other processes.  The SDD should be generated by `ShuffleMaster` by design, but the special `UnknownShuffleDeploymentDescriptor` is generated only in the case of `LocationType.Unknown` which is not via `ShuffleMaster`.

Reply:
ok, I think I see the problem now, thanks for explanation. I will put my thoughts in other order :)

2. 

During the design, I thought `ShuffleDeploymentDescriptor` was supposed to contain shuffle specific info generated by `ShuffleMaster` as a central point and used eventually by `ShuffleService` in producer and consumer Task to setup readers/writers. 

The example could be some partition identification or connection inside external shuffle system. The existing connection id/location is also an example of it for the existing netty stack, but might be not relevant for other shuffle systems.

For example, let's say the partition is stored remotely (not in producer), the batch job is restored and some the partition is finished, we do not even need to deploy the producer, just connect the consumer to the existing 'done' external partition, then the existing connection id does not make sense, the consumer needs some kind of internal shuffle id of the partition.

That is why I thought: PSD(ProducerResourceId,ProducerConnection,...) -> `ShuffleMaster` -> SDD(Internal) -> ICDD(SDD) -> Task -> ICDD,ConsumerResourceId -> `ShuffleService` -> InputGate -> read records. 

I think even `ShuffleService` itself can decide what to do with ProducerResourceId/ConsumerResourceId and calculate internally LocationType in case of existing netty. For other shuffle services, LocationType might be not relevant (like external partition), then maybe ICDD=SDD=PartitionInfo and we could leave only one of them, not sure.

I thought of `UnknownShuffleDeploymentDescriptor` as a replacement of `LocationType.Unknown\ConnectionId=null` based on the above arguments. It is just a singleton stub to signal that SDD will be updated later with the `sendUpdatePartitionInfoRpcCall` in case of lazy scheduling. True, it is not generated by `ShuffleMaster`, what could be an alternative for this approach?

1. 

In case of eager deployment (lazyScheduling=false), currently, we can already deploy the consumer when the slot is assigned to the producer but its deployment has not started yet and we planned to generate the SDD during producer deployment. If we agree on 2., it seems that we need SDD for consumer to consume and it has to be known.

Thinking more about `ShuffleMaster` interface, depending on its nature, it might be an asynchronous API like registering and talking to an external system. This means that ideally its partition register method should return a `CompletableFuture<SDD>`. 

Then the producer execution life cycle should be: created -> scheduled -> slot assigned -> register partition (get and cache SDD) -> deploying (generate TDD with previously acquired SDD). Everything happening on the main thread of Job Master. The consumer has to be deployed not after producer slot is assigned but after partition is registered in eager scheduling. In lazy scheduling, we have the `sendUpdatePartitionInfoRpcCall` to send SDD later.

 I would suggest we do the partitions registering and SDD caching in `allocateAndAssignSlotForExecution`, right after slot assignment: 
```
return logicalSlotFuture
  .thenApply(..tryAssignResource..)
  .thenCompose(..ShuffleMaster.register(PSD), cache SDDs..);
```

Reply:
If I understand correctly, the above comments can be summarized  into three points:

1.  `LocationType` can be decided by `ShuffleService` by comparing `ResourceID` between producer and consumer, then ICDD, SDD, `PartitionInfo` might be totally equivalent and we might need  retain only SDD to replace the other two classes.

2. `registerPartition` and `cachePartition` are triggered in `allocateAndAssignSlotForExecution` and return future, then we can confirm during deploying the consumer in eager mode, we can always get corresponding registered/cached SDD of producers.

3. `UnknownShuffleDeploymentDescriptor` should also be introduced for lazy deployment mode to indicate there would be updated SDD later.

I agree with the above points, especially for the point 2 which solves my previous concern naturally.

Reply:
Considering the point 1, I think ICDD might include [`SDD`, `ResultPartitionID`, `consumerResourceID`]. 
The `PartitionInfo` might be replaced by `ICDD` directly because the current only different field `IntermediateDataSetID` is only used for finding the proper `SingleInputGate` during `updatePartitionInfo` in task. Then we can use the `ResultPartitionID` in ICDD to find the proper `SingleInputGate` instead. I want to refactor this process to remove `PartitionInfo` first in a separate PR. What do you think?